### DISCUSSION TITLE
Minimal Scala support for custom scales

### DISCUSSION THREAD

**Discussion by nikodemus** [SynthstromAudible/DelugeFirmware#824] [Ideas]
There are a couple of extant discusssion relating to custom scales that I could find.

- https://github.com/SynthstromAudible/DelugeFirmware/discussions/616 : focuses on microtonality, but has also thoughts on some of the UX elements.
- https://github.com/SynthstromAudible/DelugeFirmware/discussions/759 : focuses on UI elements of programming new scales on the device.
- https://github.com/SynthstromAudible/DelugeFirmware/issues/25 : alternate tunings, obviously touching on scales as well.
- https://github.com/SynthstromAudible/DelugeFirmware/discussions/625 : discusses support for accidentals, plus the related (stale, wip) PR: https://github.com/SynthstromAudible/DelugeFirmware/pull/102

My idea is to just add minimal support for Scala files. See: https://huygens-fokker.org/scala/scl_format.html

Minimal feature set I'm thinking about:

- Folder SCALES is scanned on startup.
- Scale files found there are added as available scales to the SHIFT+SCALE button cycling. Name of the scale is taken from the filename.
- All note values are rounded to nearest semitone, which is used as the name of the note. (In absence of microtonal support.)
- Scales with more than 12 notes are either ignored, or truncated to first 12 notes. (As long as the engine uses a 12-note lookup table.)
- When transposing a clip with notes other than the root to a different scale using the SCALE-button, only cycle through scales with the same number of notes. (In absence of support for accidental notes.)

Later enchancements could include:

- Saving scales created on the device as Scala files for future use.
- Community-features submenu which allows enabling and disabling specific scales. (If you never use Locrian, why keep it in the rotation...)
- Pressing SCALE+SELECT to open a scale browser on the screen. 

I could potentially work on this myself, but I haven't looked at the Deluge codebase at all yet - so no promises.

Biggest issue I foresee is how currently hitting shift+scale button transposes to the new scale. Simplest solution would be to disallow transposing to scales with a different number of notes. 

(Since Deluge supports chromatic scale as well I'm assuming that there aren't any deep assumptions about 7 notes per scale or diatonic scales hiding anywhere...)

EDITED to:
- note the 12-note limitation I learned about while looking at other issues.
- fix my mistaken assuptiom about accidentals, link to appropriate discussion.

**Comment by samfrantz** [comment]
I do not yet own a synthstrom deluge but I applaud this community
An embedded software engineer myself, I'll be poking around the stack :)
I like all of Nico's ideas, my enhancement after all of his is to allow scale change events within the sequence - not necessarily changing intonation, but that could be cool also
Aim high 👍

**Comment by acook** [comment]
User scales would be really nice, and this way would open up a huge number of scales to us Deluge owners!

**Comment by vlork** [comment]
Another +1 for micro tuning and basic support of the Scala standard and MIDI synths that support scala. 

As far as I can tell, the standard MIDI 1.0 implementation works like follows:

1. Load Scala definition file, either text (.scl) or MIDI tuning file (.mid).
2. Definition file contains a list of microtonal intervals and the scale length. A root note (commonly C4) is assumed or specified elsewhere.
3. Set Deluge scale to name in the Scala definition file. 
4. Set Deluge scale length to the scale length in the Scala definition file. 
5. Optionally send scale definition to selected MIDI synths that support Scala tuning too. 
6. When notes are played from MIDI clips, regular MIDI notes are used. No pitch bend or other magic needed. The target synth is responsible for tuning.
7. When notes are played on Deluge’s internal synths, pitch is calculated from the root note using a microtuning lookup table rather than assuming 100 cents per semitone.

**Reply by seangoodvibes** [reply]
> Another +1 for micro tuning and basic support of the Scala standard and MIDI synths that support scala.
> 
> As far as I can tell, the standard MIDI 1.0 implementation works like follows:
> 
> 1. Load Scala definition file, either text (.scl) or MIDI tuning file (.mid).
> 2. Definition file contains a list of microtonal intervals and the scale length. A root note (commonly C4) is assumed or specified elsewhere.
> 3. Set Deluge scale to name in the Scala definition file.
> 4. Set Deluge scale length to the scale length in the Scala definition file.
> 5. Optionally send scale definition to selected MIDI synths that support Scala tuning too.
> 6. When notes are played from MIDI clips, regular MIDI notes are used. No pitch bend or other magic needed. The target synth is responsible for tuning.
> 7. When notes are played on Deluge’s internal synths, pitch is calculated from the root note using a microtuning lookup table rather than assuming 100 cents per semitone.

You can checkout this work in progress and provide feedback :)!

https://github.com/SynthstromAudible/DelugeFirmware/pull/3777

**Reply by thejesiah** [reply]
> You can checkout this work in progress and provide feedback :)!
> 
> #3777

So exciting coming to the forum to see if microtonal scales/Scala has been discussed and see this! Looks like it's getting really close!  as always, thank you all so much
